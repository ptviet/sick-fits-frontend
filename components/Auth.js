import React, { Component } from 'react';
import Router from 'next/router';
import styled from 'styled-components';
import User from './User';
import SignUp from './SignUp';
import SignIn from './SignIn';
import RequestResetPassword from './RequestResetPassword';

const Columns = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
  grid-gap: 20px;
`;

class Auth extends Component {
  render() {
    return (
      <User>
        {({ data: { me } }) => {
          if (me)
            Router.push({
              pathname: '/'
            });
          return (
            <Columns>
              <SignUp />
              <SignIn />
              <RequestResetPassword />
            </Columns>
          );
        }}
      </User>
    );
  }
}

export default Auth;
