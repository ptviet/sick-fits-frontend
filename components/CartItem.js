import React from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3 {
    margin: 0;
  }
`;

const CartItem = ({ cartItem }) => {
  return (
    <CartItemStyles>
      {cartItem.item ? (
        <>
          <img
            src={cartItem.item.image}
            alt={cartItem.item.title}
            width="100"
          />
          <div className="cart-item-details">
            <h3>{cartItem.item.title}</h3>
            <p>
              <em>
                {cartItem.quantity} &times; {formatMoney(cartItem.item.price)}
                {' (each)'}
              </em>
              {' = '}
              {formatMoney(cartItem.item.price * cartItem.quantity)}
            </p>
          </div>
        </>
      ) : (
        <p>Item has been removed by the seller.</p>
      )}
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  );
};

export default CartItem;
