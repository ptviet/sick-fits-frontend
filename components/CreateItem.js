import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import gql from 'graphql-tag';
import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $image: String
    $largeImage: String
    $price: Int!
  ) {
    createItem(
      title: $title
      description: $description
      image: $image
      largeImage: $largeImage
      price: $price
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0
  };

  onChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(Number(value)) : value;
    this.setState({ [name]: val });
  };

  data = null;
  onFileChange = e => {
    e.preventDefault();
    const files = e.target.files;
    this.data = new FormData();
    this.data.append('file', files[0]);
    this.data.append('upload_preset', 'sickfits');
    this.previewImage(files[0]);
  };

  previewImage(file) {
    if (file != null || file != undefined) {
      var oFReader = new FileReader();
      oFReader.readAsDataURL(file);

      oFReader.onload = oFREvent => {
        document.getElementById('uploadPreview').src = oFREvent.target.result;
        document.getElementById('uploadPreview').alt = 'Upload Preview';
        document.getElementById('uploadPreview').style = 'height: 100px';
        document.getElementById('uploadPreview').hidden = false;
      };
    } else {
      document.getElementById('uploadPreview').hidden = true;
      this.data = null;
    }
  }

  onSubmit = async (e, createItemMutation) => {
    e.preventDefault();
    if (this.data != null) {
      const uploadRes = await fetch(
        'https://api.cloudinary.com/v1_1/ptviet/image/upload',
        {
          method: 'POST',
          body: this.data
        }
      );
      const file = await uploadRes.json();
      await this.setState({
        image: file.secure_url,
        largeImage: file.eager[0].secure_url
      });
      const res = await createItemMutation();
      Router.push({
        pathname: '/item',
        query: { id: res.data.createItem.id }
      });
    }
  };

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form onSubmit={e => this.onSubmit(e, createItem)}>
            <ErrorMessage error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <img id="uploadPreview" />
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.onFileChange}
                />
              </label>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  value={this.state.title}
                  required
                  onChange={this.onChange}
                />
              </label>
              <label htmlFor="price">
                Price (in cent)
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  value={this.state.price}
                  required
                  onChange={this.onChange}
                />
              </label>
              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Description"
                  value={this.state.description}
                  required
                  onChange={this.onChange}
                />
              </label>
              <button type="submit">Submit{loading && 'ting'}</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
