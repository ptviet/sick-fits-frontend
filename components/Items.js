import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Item from './Item';
import Pagination from './Pagination';
import ErrorMessage from './ErrorMessage';
import { perPage } from '../config';

const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($first: Int = ${perPage}, $skip: Int = 0 ) {
    items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
      id
      title
      description
      price
      image
      largeImage
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(330px, 1fr));
  grid-gap: 20px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;

class Items extends Component {
  render() {
    return (
      <Center>
        <Pagination page={this.props.page} />
        <Query
          query={ALL_ITEMS_QUERY}
          // fetchPolicy='network-only'
          variables={{
            skip: this.props.page * perPage - perPage
          }}
        >
          {({ data, error, loading }) => {
            if (loading) return <p>Loading...</p>;
            if (error) return <ErrorMessage error={error} />;
            return (
              <ItemsList>
                {data.items.map(item => (
                  <Item key={item.id} item={item} />
                ))}
              </ItemsList>
            );
          }}
        </Query>
        <Pagination page={this.props.page} />
      </Center>
    );
  }
}

export default Items;
export { ALL_ITEMS_QUERY };
