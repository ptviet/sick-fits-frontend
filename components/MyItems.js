import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import User from './User';
import Item from './Item';
import ErrorMessage from './ErrorMessage';

const MY_ITEMS_QUERY = gql`
  query MY_ITEMS_QUERY($id: ID!) {
    items(where: { user: { id: $id } }, orderBy: createdAt_DESC) {
      id
      title
      description
      price
      image
      largeImage
    }
  }
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;

class MyItems extends Component {
  render() {
    return (
      <User>
        {({ data: { me } }) => {
          if (!me) return <p>Please login.</p>;
          return (
            <Query query={MY_ITEMS_QUERY} variables={{ id: me.id }}>
              {({ data, error, loading }) => {
                if (loading) return <p>Loading...</p>;
                if (error) return <ErrorMessage error={error} />;
                return (
                  <>
                    <h2>Your item(s)</h2>
                    {data.items.length === 0 && !loading && (
                      <p>You don't have any item.</p>
                    )}
                    <ItemsList>
                      {data.items.map(item => (
                        <Item key={item.id} item={item} />
                      ))}
                    </ItemsList>
                  </>
                );
              }}
            </Query>
          );
        }}
      </User>
    );
  }
}

export default MyItems;
