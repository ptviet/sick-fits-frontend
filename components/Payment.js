import React, { Component } from 'react';
import StripeCheckout from 'react-stripe-checkout';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import NProgress from 'nprogress';
import User, { CURRENT_USER_QUERY } from './User';
import calcTotalPrice from '../lib/calcTotalPrice';
import { paymentIcon, STRIPE_PUBLISHABLE_KEY } from '../config';

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    createOrder(token: $token) {
      id
      total
      charge
      items {
        id
        title
      }
    }
  }
`;

class Payment extends Component {
  totalItems = cart =>
    cart.reduce((total, cartItem) => total + cartItem.quantity, 0);

  onToken = async (res, createOrderMutation) => {
    NProgress.start();
    if (res.id) {
      const order = await createOrderMutation({
        variables: {
          token: res.id
        }
      }).catch(error => alert(error.message.replace('GraphQL error: ', '')));
      // Redirect
      Router.push({
        pathname: '/order',
        query: {
          id: order.data.createOrder.id
        }
      });
    }
  };

  render() {
    return (
      <User>
        {({ data: { me }, loading }) => {
          if (loading) return <p>Loading...</p>;
          return (
            <Mutation
              mutation={CREATE_ORDER_MUTATION}
              refetchQueries={[{ query: CURRENT_USER_QUERY }]}
            >
              {createOrder => (
                <StripeCheckout
                  amount={calcTotalPrice(me.cart)}
                  name="Sick Fits"
                  description={`Order of ${this.totalItems(me.cart)} item${
                    this.totalItems(me.cart) > 1 ? 's' : ''
                  }`}
                  image={paymentIcon}
                  stripeKey={STRIPE_PUBLISHABLE_KEY}
                  currency="AUD"
                  email={me.email}
                  token={res => this.onToken(res, createOrder)}
                >
                  {this.props.children}
                </StripeCheckout>
              )}
            </Mutation>
          );
        }}
      </User>
    );
  }
}

export default Payment;
export { CREATE_ORDER_MUTATION };
