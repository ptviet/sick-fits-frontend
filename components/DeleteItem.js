import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  update = (cache, payload) => {
    // Update the cache on the client to match the server
    // Read the cache to get the item
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // Filter the deleted item out of the page
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    );
    // Put the items back
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  };

  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {deleteItem => (
          <button
            onClick={() => {
              if (confirm('Are you sure you want to delete this item?')) {
                deleteItem().catch(error =>
                  alert(error.message.replace('GraphQL error: ', ''))
                );
              }
            }}
          >
            {this.props.children}
          </button>
        )}
      </Mutation>
    );
  }
}

export default DeleteItem;
