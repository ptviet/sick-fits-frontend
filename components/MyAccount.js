import React, { Component } from 'react';
import Head from 'next/head';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import MyItems from './MyItems';

const MY_INFO_QUERY = gql`
  query MY_INFO_QUERY {
    me {
      id
      email
      name
    }
  }
`;

const UPDATE_PASSWORD_MUTATION = gql`
  mutation UPDATE_PASSWORD_MUTATION(
    $currentPassword: String!
    $password: String!
    $confirmPassword: String!
  ) {
    updatePassword(
      currentPassword: $currentPassword
      password: $password
      confirmPassword: $confirmPassword
    ) {
      id
      email
      name
    }
  }
`;

class MyAccount extends Component {
  state = {
    currentPassword: '',
    password: '',
    confirmPassword: ''
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onUpdatePassword = async (e, updatePasswordMutation) => {
    e.preventDefault();
    if (this.state.password !== this.state.confirmPassword) {
      alert("Your passwords don't match!");
    }
    if (
      this.state.password === this.state.confirmPassword &&
      this.state.currentPassword.length > 0
    ) {
      await updatePasswordMutation();
      this.setState({
        currentPassword: '',
        password: '',
        confirmPassword: ''
      });
      alert('Your password has been changed!');
    }
  };

  render() {
    return (
      <>
        <Head>
          <title>Sick Fits | My Account</title>
        </Head>
        <Query query={MY_INFO_QUERY}>
          {({ data: { me }, error, loading }) => {
            if (loading) return <p>Loading...</p>;
            if (error) return <ErrorMessage error={error} />;
            if (!me) return <p>Please login.</p>;
            return (
              <Mutation
                mutation={UPDATE_PASSWORD_MUTATION}
                variables={{
                  currentPassword: this.state.currentPassword,
                  password: this.state.password,
                  confirmPassword: this.state.confirmPassword
                }}
              >
                {(updatePassword, { error, loading }) => (
                  <Form
                    method="post"
                    onSubmit={e => this.onUpdatePassword(e, updatePassword)}
                  >
                    <fieldset disabled={loading} aria-busy={loading}>
                      <h2>Account Info</h2>
                      <ErrorMessage error={error} />
                      <label htmlFor="email">
                        Email
                        <input
                          disabled
                          type="email"
                          name="email"
                          value={me.email}
                        />
                      </label>
                      <label htmlFor="name">
                        Name
                        <input
                          disabled
                          type="text"
                          name="name"
                          value={me.name}
                        />
                      </label>
                      <label htmlFor="password">
                        Current Password
                        <input
                          type="password"
                          name="currentPassword"
                          placeholder="Current Password"
                          value={this.state.currentPassword}
                          onChange={this.onChange}
                        />
                      </label>
                      <label htmlFor="password">
                        New Password
                        <input
                          type="password"
                          name="password"
                          placeholder="Password"
                          value={this.state.password}
                          onChange={this.onChange}
                        />
                      </label>
                      <label htmlFor="password">
                        Confirm Password
                        <input
                          type="password"
                          name="confirmPassword"
                          placeholder="Confirm Password"
                          value={this.state.confirmPassword}
                          onChange={this.onChange}
                        />
                      </label>
                      <button
                        type="submit"
                        disabled={
                          this.state.currentPassword.length === 0 ||
                          this.state.password.length === 0 ||
                          this.state.confirmPassword.length === 0
                        }
                      >
                        Change Password
                      </button>
                    </fieldset>
                  </Form>
                )}
              </Mutation>
            );
          }}
        </Query>
        <MyItems />
      </>
    );
  }
}

export default MyAccount;
