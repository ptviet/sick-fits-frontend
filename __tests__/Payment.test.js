import { MockedProvider } from 'react-apollo/test-utils';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Router from 'next/router';
import NProgress from 'nprogress';
import wait from 'waait';
import Payment from '../components/Payment';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeCartItem, fakeUser, fakeOrder } from '../lib/testUtils';

Router.router = { push: jest.fn() };
NProgress.start = jest.fn();

const user = fakeUser();
const cartItem = fakeCartItem();
const order = fakeOrder();

const createOrderMock = jest.fn().mockResolvedValue({
  data: {
    createOrder: {
      ...order
    }
  }
});

const mocks = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: {
        me: { ...user, cart: [cartItem] }
      }
    }
  }
];

describe('<Payment />', () => {
  let wrapper;
  it('should render and match snapshot', async () => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Payment />
      </MockedProvider>
    );
    await wait();
    wrapper.update();

    const StripeCheckoutButton = wrapper.find('StripeCheckout');
    expect(toJson(StripeCheckoutButton)).toMatchSnapshot();
  });

  it('should create an order onToken', async () => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Payment />
      </MockedProvider>
    );
    await wait();
    wrapper.update();

    const component = wrapper.find('Payment').instance();

    // Call onToken method
    component.onToken({ id: 'abc123' }, createOrderMock);
    expect(createOrderMock).toHaveBeenCalled();
    expect(createOrderMock).toHaveBeenCalledWith({
      variables: { token: 'abc123' }
    });
  });

  it('should turn the progress bar on', async () => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Payment />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const component = wrapper.find('Payment').instance();

    // Call onToken method
    component.onToken({ id: 'abc123' }, createOrderMock);

    expect(NProgress.start).toHaveBeenCalled();
  });

  it('should route to the order page when completed', async () => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Payment />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const component = wrapper.find('Payment').instance();

    // Call onToken method
    component.onToken({ id: 'abc123' }, createOrderMock);
    expect(Router.router.push).toHaveBeenCalled();
  });
});
