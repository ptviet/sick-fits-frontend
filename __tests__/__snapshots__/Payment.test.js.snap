// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Payment /> should render and match snapshot 1`] = `null`;
